src/main/java/com/twcable/grabbit/tools/cli/GrabbitCli.java

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright 2014-2016 Time Warner Cable, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.twcable.grabbit.tools.cli;
+
+import com.twcable.grabbit.tools.environment.Environment;
+import com.twcable.grabbit.tools.environment.HostInfo;
+import com.twcable.grabbit.tools.environment.MapSourcedEnvironment;
+import com.twcable.grabbit.tools.jobstarter.HostAndJobId;
+import com.twcable.grabbit.tools.jobstarter.JobStarter;
+import com.twcable.grabbit.tools.jobstarter.JobsConfigFileReader;
+import com.twcable.grabbit.tools.monitor.JobStatusCache;
+import com.twcable.grabbit.tools.monitor.PollingJobMonitor;
+import com.twcable.grabbit.tools.monitor.SimpleFileJobStatusCache;
+import com.twcable.grabbit.tools.monitor.event.MonitoringEvent;
+import com.twcable.grabbit.tools.util.Either;
+import com.twcable.grabbit.tools.util.Utils;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.WorkQueueProcessor;
+import reactor.core.subscriber.Subscribers;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CountDownLatch;
+import java.util.stream.Collectors;
+
+import static com.twcable.grabbit.tools.monitor.JobState.RUNNING;
+
+/**
+ * The command line entry point.
+ */
+@SuppressWarnings({"WeakerAccess", "RedundantTypeArguments"})
+public class GrabbitCli {
+    public static final String DEFAULT_JOB_STATUS_CACHE_FILENAME = "grabbitIds.out";
+
+
+    @SuppressWarnings({"PMD.UseVarargs", "PMD.SystemPrintln"})
+    public static void main(String[] args) throws IOException, InterruptedException {
+        final Either<String, CliOptions> cliOptions = CliOptions.create(args);
+
+        if (cliOptions.isRight()) {
```

Comment:
My first inclination was to do different CliOptions subclasses, like `StartCliOptions`, `MonitorCliOptions`, `StartAndMonitorCliOptions`, `InvalidCliOptions`. The problem, as you noted, is that they are disjoint types, which Java really doesn't handle well. Unlike, for example, Haskell.

Speaking of other languages, that's where I got `Either` from. Specifically, it's a simplified version of [Javaslang's Either type](https://static.javadoc.io/io.javaslang/javaslang/2.0.2/javaslang/control/Either.html), which it took from Scala. But the concept is common in functional programming since you want to propagate errors and the like through the system without relying on stack-popping behavior except for truly "exceptional" cases. If you ever do GoLang, it's also a standard idiom: most functions return a tuple, where the left value is a potential error.

I was going to use it more, but then when I added the Reactive-Streams stuff that reduced/changed the need. So as this continues to evolve I may or may not keep it.